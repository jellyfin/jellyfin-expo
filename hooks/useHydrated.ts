/**
 * Copyright (c) 2025 Jellyfin Contributors
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

import { useEffect, useState } from 'react';

import { useDownloadStore } from '../stores/DownloadStore';
import { useRootStore } from '../stores/RootStore';
import { useServerStore } from '../stores/ServerStore';
import { useSettingStore } from '../stores/SettingStore';

export const useIsHydrated = () => {
	const [ isHydrated, setIsHydrated ] = useState(false);
	const [ isRootStoreHydrated, setIsRootStoreHydrated ] = useState(false);
	const [ isDownloadStoreHydrated, setIsDownloadStoreHydrated ] = useState(false);
	const [ isServerStoreHydrated, setIsServerStoreHydrated ] = useState(false);
	const [ isSettingStoreHydrated, setIsSettingStoreHydrated ] = useState(false);

	useEffect(() => {
		// Root Store
		const unsubRootFinishHydration = useRootStore.persist.onFinishHydration(() => setIsRootStoreHydrated(true));
		setIsRootStoreHydrated(useRootStore.persist.hasHydrated());
		// Download Store
		const unsubDownloadFinishHydration = useDownloadStore.persist.onFinishHydration(() => setIsDownloadStoreHydrated(true));
		setIsDownloadStoreHydrated(useDownloadStore.persist.hasHydrated());
		// Server Store
		const unsubServerFinishHydration = useServerStore.persist.onFinishHydration((state) => {
			console.debug('>>> hydrated state', state);
			setIsServerStoreHydrated(true);
		});
		setIsServerStoreHydrated(useServerStore.persist.hasHydrated());
		// Settings Store
		const unsubSettingFinishHydration = useSettingStore.persist.onFinishHydration(() => setIsSettingStoreHydrated(true));
		setIsSettingStoreHydrated(useSettingStore.persist.hasHydrated());

		return () => {
			unsubRootFinishHydration();
			unsubDownloadFinishHydration();
			unsubServerFinishHydration();
			unsubSettingFinishHydration();
		};
	}, []);

	useEffect(() => {
		console.debug('Hydration state', isRootStoreHydrated, isDownloadStoreHydrated, isServerStoreHydrated, isSettingStoreHydrated);
		if (isRootStoreHydrated && isDownloadStoreHydrated && isServerStoreHydrated && isSettingStoreHydrated) {
			setIsHydrated(true);
		}
	}, [ isDownloadStoreHydrated, isRootStoreHydrated, isServerStoreHydrated, isSettingStoreHydrated ]);

	return isHydrated;
};
