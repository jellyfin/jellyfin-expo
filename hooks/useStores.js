/**
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */
import { useDownloadStore } from '../stores/DownloadStore';
import { useMediaStore } from '../stores/MediaStore';
import { useRootStore } from '../stores/RootStore';
import { useServerStore } from '../stores/ServerStore';
import { useSettingStore } from '../stores/SettingStore';

// Compatibility for zustand conversion
export const useStores = () => ({
	rootStore: useRootStore(),
	downloadStore: useDownloadStore(),
	serverStore: useServerStore(),
	mediaStore: useMediaStore(),
	settingStore: useSettingStore()
});
