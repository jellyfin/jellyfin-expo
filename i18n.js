/**
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */
import * as Localization from 'expo-localization';
import i18next from 'i18next';
import { initReactI18next } from 'react-i18next';
import { I18nManager } from 'react-native';

import ar from './langs/ar.json';
import be from './langs/be.json';
import bg from './langs/bg.json';
import ca from './langs/ca.json';
import cs from './langs/cs.json';
import cy from './langs/cy.json';
import da from './langs/da.json';
import de from './langs/de.json';
import el from './langs/el.json';
import en from './langs/en.json';
import eo from './langs/eo.json';
import es from './langs/es.json';
import es_419 from './langs/es_419.json';
import es_AR from './langs/es_AR.json';
import et from './langs/et.json';
import fa from './langs/fa.json';
import fi from './langs/fi.json';
import fil from './langs/fil.json';
import fr from './langs/fr.json';
import ga from './langs/ga.json';
import he from './langs/he.json';
import hr from './langs/hr.json';
import hu from './langs/hu.json';
import hy from './langs/hy.json';
import id from './langs/id.json';
import is from './langs/is.json';
import it from './langs/it.json';
import ja from './langs/ja.json';
import kk from './langs/kk.json';
import ko from './langs/ko.json';
import lt from './langs/lt.json';
import lv from './langs/lv.json';
import mn from './langs/mn.json';
import ms from './langs/ms.json';
import mt from './langs/mt.json';
import nb_NO from './langs/nb_NO.json';
import nds from './langs/nds.json';
import nl from './langs/nl.json';
import pl from './langs/pl.json';
import pt from './langs/pt.json';
import pt_BR from './langs/pt_BR.json';
import pt_PT from './langs/pt_PT.json';
import ro from './langs/ro.json';
import ru from './langs/ru.json';
import sk from './langs/sk.json';
import sl from './langs/sl.json';
import sq from './langs/sq.json';
import sr from './langs/sr.json';
import sv from './langs/sv.json';
import ta from './langs/ta.json';
import tr from './langs/tr.json';
import uk from './langs/uk.json';
import vi from './langs/vi.json';
import zh_Hans from './langs/zh_Hans.json';
import zh_Hant from './langs/zh_Hant.json';

export const resources = {
	en: { translation: en },
	ar: { translation: ar },
	be: { translation: be },
	bg: { translation: bg },
	ca: { translation: ca },
	cs: { translation: cs },
	cy: { translation: cy },
	da: { translation: da },
	de: { translation: de },
	el: { translation: el },
	eo: { translation: eo },
	es: { translation: es },
	'es-419': { translation: es_419 },
	'es-AR': { translation: es_AR },
	et: { translation: et },
	fa: { translation: fa },
	fi: { translation: fi },
	fil: { translation: fil },
	fr: { translation: fr },
	ga: { translation: ga },
	he: { translation: he },
	hr: { translation: hr },
	hu: { translation: hu },
	hy: { translation: hy },
	id: { translation: id },
	is: { translation: is },
	it: { translation: it },
	ja: { translation: ja },
	kk: { translation: kk },
	ko: { translation: ko },
	lt: { translation: lt },
	lv: { translation: lv },
	mn: { translation: mn },
	ms: { translation: ms },
	mt: { translation: mt },
	'nb-NO': { translation: nb_NO },
	nds: { translation: nds },
	nl: { translation: nl },
	pl: { translation: pl },
	pt: { translation: pt },
	'pt-BR': { translation: pt_BR },
	'pt-PT': { translation: pt_PT },
	ro: { translation: ro },
	ru: { translation: ru },
	sk: { translation: sk },
	sl: { translation: sl },
	sq: { translation: sq },
	sr: { translation: sr },
	sv: { translation: sv },
	ta: { translation: ta },
	tr: { translation: tr },
	uk: { translation: uk },
	vi: { translation: vi },
	'zh-Hans': { translation: zh_Hans },
	'zh-Hant': { translation: zh_Hant }
};

// Ensure RTL layout is enabled for RTL locales
I18nManager.forceRTL(Localization.isRTL);

i18next
	.use(initReactI18next)
	.init({
		// debug: true,
		fallbackLng: 'en',
		// Use the fallback language for empty strings
		returnEmptyString: false,
		lng: Localization.locale,
		interpolation: {
			escapeValue: false
		},
		resources
	});

// Export i18next instance for use in tests
export default i18next;
