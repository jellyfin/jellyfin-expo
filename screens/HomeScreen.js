/**
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */
import { useFocusEffect, useNavigation } from '@react-navigation/native';
import React, { useCallback, useContext, useEffect, useRef, useState } from 'react';
import { useTranslation } from 'react-i18next';
import { BackHandler, Platform, StyleSheet, View } from 'react-native';
import { ThemeContext } from 'react-native-elements';
import { SafeAreaView, useSafeAreaInsets } from 'react-native-safe-area-context';

import AudioPlayer from '../components/AudioPlayer';
import ErrorView from '../components/ErrorView';
import NativeShellWebView from '../components/NativeShellWebView';
import VideoPlayer from '../components/VideoPlayer';
import Colors from '../constants/Colors';
import MediaTypes from '../constants/MediaTypes';
import Screens from '../constants/Screens';
import { useStores } from '../hooks/useStores';
import { getIconName } from '../utils/Icons';

const HomeScreen = () => {
	const { rootStore, serverStore, mediaStore, settingStore } = useStores();
	const navigation = useNavigation();
	const { t } = useTranslation();
	const insets = useSafeAreaInsets();
	const { theme } = useContext(ThemeContext);

	const [ isLoading, setIsLoading ] = useState(true);
	const [ httpErrorStatus, setHttpErrorStatus ] = useState(null);

	const webview = useRef(null);

	useEffect(() => {
		// Pressing the Home tab when it is already active navigates to home screen in webview
		navigation.getParent()?.addListener('tabPress', e => {
			if (navigation.isFocused()) {
				// Prevent default behavior
				e.preventDefault();
				// Call the web router to navigate home
				webview.current?.injectJavaScript('window.ExpoRouterShim && window.ExpoRouterShim.home();');
			}
		});
	}, []);

	useFocusEffect(
		useCallback(() => {
			const onBackPress = () => {
				webview.current?.injectJavaScript('window.ExpoRouterShim && window.ExpoRouterShim.back();');
				return true;
			};

			BackHandler.addEventListener('hardwareBackPress', onBackPress);

			return () => BackHandler.removeEventListener('hardwareBackPress', onBackPress);
		}, [ webview ])
	);

	// Report media updates to the audio/video plugin
	useEffect(() => {
		if (!mediaStore.isLocalFile) {
			const status = {
				didPlayerCloseManually: rootStore.didPlayerCloseManually,
				uri: mediaStore.uri,
				isFinished: mediaStore.isFinished,
				isPlaying: mediaStore.isPlaying,
				positionTicks: mediaStore.positionTicks,
				positionMillis: mediaStore.positionMillis
			};

			if (mediaStore.type === MediaTypes.Audio) {
				webview.current?.injectJavaScript(`window.ExpoAudioPlayer && window.ExpoAudioPlayer._reportStatus(${JSON.stringify(status)});`);
			} else if (mediaStore.type === MediaTypes.Video) {
				webview.current?.injectJavaScript(`window.ExpoVideoPlayer && window.ExpoVideoPlayer._reportStatus(${JSON.stringify(status)});`);
			}
		}
	}, [
		mediaStore.type,
		mediaStore.uri,
		mediaStore.isFinished,
		mediaStore.isLocalFile,
		mediaStore.isPlaying,
		mediaStore.positionTicks
	]);

	// Clear the error state when the active server changes
	useEffect(() => {
		setIsLoading(true);
	}, [ settingStore.activeServer ]);

	useEffect(() => {
		if (rootStore.isReloadRequired) {
			webview.current?.reload();
			rootStore.set({ isReloadRequired: false });
		}
	}, [ rootStore.isReloadRequired ]);

	useEffect(() => {
		if (httpErrorStatus) {
			const errorCode = httpErrorStatus.description || httpErrorStatus.statusCode;
			navigation.replace(Screens.ErrorScreen, {
				icon: {
					name: 'cloud-off',
					type: 'material'
				},
				heading: t([ `home.errors.${errorCode}.heading`, 'home.errors.http.heading' ]),
				message: t([ `home.errors.${errorCode}.description`, 'home.errors.http.description' ]),
				details: [
					t('home.errorCode', { errorCode }),
					t('home.errorUrl', { url: httpErrorStatus.url })
				],
				buttonIcon: {
					name: getIconName('refresh'),
					type: 'ionicon'
				},
				buttonTitle: t('home.retry')
			});
		}
	}, [ httpErrorStatus ]);

	// When not in fullscreen, the top adjustment is handled by the spacer View for iOS
	const safeAreaEdges = [ 'right', 'left' ];
	if (Platform.OS !== 'ios' || rootStore.isFullscreen) {
		safeAreaEdges.push('top');
	}
	// Bottom spacer is handled by tab bar except in fullscreen
	if (rootStore.isFullscreen) {
		safeAreaEdges.push('bottom');
	}
	// Hide webview until loaded
	const webviewStyle = (isLoading || httpErrorStatus) ? StyleSheet.compose(styles.container, styles.loading) : styles.container;

	if (!serverStore.servers || serverStore.servers.length === 0) {
		return null;
	}
	const server = serverStore.servers[settingStore.activeServer];

	return (
		<SafeAreaView
			style={{
				...styles.container,
				backgroundColor: rootStore.isFullscreen ? Colors.black : theme.colors.background
			}}
			edges={safeAreaEdges}
		>
			{Platform.OS === 'ios' && !rootStore.isFullscreen && (
				<View style={{
					backgroundColor: theme.colors.grey0,
					height: insets.top
				}} />
			)}
			{server && server.urlString ? (
				<>
					<NativeShellWebView
						ref={webview}
						style={webviewStyle}
						containerStyle={webviewStyle}
						refreshControlProps={{
							// iOS colors
							tintColor: theme.colors.grey1,
							backgroundColor: theme.colors.grey0,
							// Android colors
							colors: [ theme.colors.primary, theme.colors.secondary ],
							progressBackgroundColor: theme.colors.background
						}}
						// Error screen is displayed if loading fails
						renderError={errorCode => (
							<ErrorView
								icon={{
									name: 'cloud-off',
									type: 'material'
								}}
								heading={t([ `home.errors.${errorCode}.heading`, 'home.errors.offline.heading' ])}
								message={t([ `home.errors.${errorCode}.description`, 'home.errors.offline.description' ])}
								details={[
									t('home.errorCode', { errorCode }),
									t('home.errorUrl', { url: server.urlString })
								]}
								buttonIcon={{
									name: getIconName('refresh'),
									type: 'ionicon'
								}}
								buttonTitle={t('home.retry')}
								onPress={() => webview.current?.reload()}
							/>
						)}
						// Loading screen is displayed when refreshing
						renderLoading={() => <View style={styles.container} />}
						// Update state on loading error
						onError={({ nativeEvent: state }) => {
							console.warn('Error', state);
						}}
						onHttpError={({ nativeEvent: state }) => {
							console.warn('HTTP Error', state);
							setHttpErrorStatus(state);
						}}
						onLoadStart={() => {
							setIsLoading(true);
							setHttpErrorStatus(null);
						}}
						// Update state when loading is complete
						onLoadEnd={() => {
							setIsLoading(false);
						}}
						// Reload the webview if the process terminated in the background
						// refs: https://github.com/react-native-webview/react-native-webview/blob/1d8205af06dbb0bad0d8f208bb2a37ce5f732fd3/docs/Reference.md#oncontentprocessdidterminate
						onContentProcessDidTerminate={() => {
							webview.current?.reload();
						}}
					/>
					<AudioPlayer/>
					<VideoPlayer/>
				</>
			) : (
				<ErrorView
					heading={t('home.errors.invalidServer.heading')}
					message={t('home.errors.invalidServer.description')}
				/>
			)}
		</SafeAreaView>
	);
};

const styles = StyleSheet.create({
	container: {
		flex: 1
	},
	loading: {
		opacity: 0
	}
});

export default HomeScreen;
