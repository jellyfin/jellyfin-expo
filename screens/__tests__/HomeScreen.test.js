/**
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

import { NavigationContainer } from '@react-navigation/native';
import { render } from '@testing-library/react-native';
import React from 'react';
import { ThemeProvider } from 'react-native-elements';
import { SafeAreaProvider } from 'react-native-safe-area-context';

import '../../i18n';
import { useStores } from '../../hooks/useStores';
import HomeScreen from '../HomeScreen';

jest.mock('@react-navigation/native', () => {
	const actualNav = jest.requireActual('@react-navigation/native');
	return {
		...actualNav,
		useFocusEffect: jest.fn()
	};
});

jest.mock('../../components/NativeShellWebView', () => 'NativeShellWebView');

jest.mock('../../hooks/useStores');
useStores.mockImplementation(() => ({
	rootStore: {
	},
	settingStore: {
		activeServer: 0
	},
	mediaStore: {},
	serverStore: {
		servers: [
			{
				urlString: 'https://example.com'
			}
		]
	}
}));

describe('HomeScreen', () => {
	it('should render correctly', () => {
		const { toJSON } = render(
			<SafeAreaProvider>
				<ThemeProvider>
					<NavigationContainer>
						<HomeScreen />
					</NavigationContainer>
				</ThemeProvider>
			</SafeAreaProvider>
		);

		expect(toJSON()).toMatchSnapshot();
	});

	it('should render null when no servers are present', () => {
		useStores.mockImplementationOnce(() => ({
			rootStore: {
			},
			settingStore: {},
			mediaStore: {},
			serverStore: {}
		}));

		const { toJSON } = render(
			<SafeAreaProvider>
				<ThemeProvider>
					<NavigationContainer>
						<HomeScreen />
					</NavigationContainer>
				</ThemeProvider>
			</SafeAreaProvider>
		);

		expect(toJSON()).toMatchSnapshot();
	});

	it('should render ErrorView when invalid server exists', () => {
		useStores.mockImplementationOnce(() => ({
			rootStore: {},
			settingStore: {
				activeServer: 0
			},
			mediaStore: {},
			serverStore: {
				servers: [{}]
			}
		}));

		const { toJSON } = render(
			<SafeAreaProvider>
				<ThemeProvider>
					<NavigationContainer>
						<HomeScreen />
					</NavigationContainer>
				</ThemeProvider>
			</SafeAreaProvider>
		);

		expect(toJSON()).toMatchSnapshot();
	});
});
