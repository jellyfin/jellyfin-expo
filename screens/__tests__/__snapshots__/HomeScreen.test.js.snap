// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeScreen should render ErrorView when invalid server exists 1`] = `
<RNCSafeAreaView
  edges={
    {
      "bottom": "off",
      "left": "additive",
      "right": "additive",
      "top": "off",
    }
  }
  style={
    {
      "backgroundColor": undefined,
      "flex": 1,
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "#393e42",
        "height": 0,
      }
    }
  />
  <View
    style={
      {
        "backgroundColor": undefined,
        "bottom": 0,
        "flex": 1,
        "left": 0,
        "paddingHorizontal": 15,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <View
      style={
        {
          "flexGrow": 1,
          "justifyContent": "center",
        }
      }
    >
      <View
        style={
          {
            "overflow": "hidden",
          }
        }
      >
        <View
          replaceTheme={[Function]}
          testID="error-view-icon"
          updateTheme={[Function]}
        >
          <View
            style={
              {
                "alignItems": "center",
                "backgroundColor": "transparent",
                "justifyContent": "center",
              }
            }
          >
            <Text />
          </View>
        </View>
      </View>
      <Text
        style={
          {
            "color": "#242424",
            "fontSize": 43.714285714285715,
            "marginVertical": 40,
            "textAlign": "center",
          }
        }
        testID="error-view-heading"
      >
        Server is invalid
      </Text>
      <Text
        style={
          {
            "color": "#242424",
            "marginBottom": 40,
            "marginHorizontal": 20,
            "textAlign": "center",
          }
        }
        testID="error-view-message"
      >
        Something has happened and the saved server information is invalid. Try removing the server from Settings and add it again.
      </Text>
    </View>
    <View
      testID="error-view-details"
    />
  </View>
</RNCSafeAreaView>
`;

exports[`HomeScreen should render correctly 1`] = `
<RNCSafeAreaView
  edges={
    {
      "bottom": "off",
      "left": "additive",
      "right": "additive",
      "top": "off",
    }
  }
  style={
    {
      "backgroundColor": undefined,
      "flex": 1,
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "#393e42",
        "height": 0,
      }
    }
  />
  <NativeShellWebView
    containerStyle={
      [
        {
          "flex": 1,
        },
        {
          "opacity": 0,
        },
      ]
    }
    onContentProcessDidTerminate={[Function]}
    onError={[Function]}
    onHttpError={[Function]}
    onLoadEnd={[Function]}
    onLoadStart={[Function]}
    refreshControlProps={
      {
        "backgroundColor": "#393e42",
        "colors": [
          "#2089dc",
          "#ca71eb",
        ],
        "progressBackgroundColor": undefined,
        "tintColor": "#43484d",
      }
    }
    renderError={[Function]}
    renderLoading={[Function]}
    style={
      [
        {
          "flex": 1,
        },
        {
          "opacity": 0,
        },
      ]
    }
  />
  <View
    style={
      [
        {
          "overflow": "hidden",
          "pointerEvents": "box-none",
        },
        undefined,
      ]
    }
  >
    <ViewManagerAdapter_ExpoVideoView
      onError={[Function]}
      onFullscreenUpdate={[Function]}
      onLoad={[Function]}
      onLoadStart={[Function]}
      onReadyForDisplay={[Function]}
      onStatusUpdate={[Function]}
      resizeMode="mock"
      status={{}}
      style={
        [
          {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          },
          undefined,
        ]
      }
      useNativeControls={true}
      videoStyle={
        [
          {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          },
          undefined,
        ]
      }
    />
    <Image
      source={
        {
          "uri": undefined,
        }
      }
      style={
        [
          {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "resizeMode": "contain",
            "right": 0,
            "top": 0,
          },
          undefined,
        ]
      }
    />
  </View>
</RNCSafeAreaView>
`;

exports[`HomeScreen should render null when no servers are present 1`] = `null`;
